utils/log_parser.py

```diff
@@ -15,136 +15,214 @@
+            oldNode: 'Node',
+            oldResNo: int,
+            newNode: 'Node',
+            newResNo: int) -> None:
+        """Replace one operand with another one.
+
+        Args:
+            oldNode: Node. Old operand node.
+            oldResNo: int. Old operand result number.
+            newNode: 'Node'. New operand node.
+            newResNo: int. New operand result number.
+        """
+
+        try:
+            self.inputs_.remove((oldNode, oldResNo))
+        except ValueError:
+            print("Removed input value must already exist in the node's input list. ")
+        self.inputs_.append((newNode, newResNo))
+
+    def set_scope_of_creation(self, creationScopeName) -> None:
         self.creationScopeName_ = creationScopeName
 
 
 class DottyPrinter:
+    """A class for generating the dotty graph file
+
+    Public attributes:
+        vertices_: List[str]. Vertices in the dotty file.
+        edges_: List[str]. Edges in the dotty file.
+        uniqueVertexMap_: Dict[Node, int]. A map for node with their unique index.
+        uniqueVertexNo_: int. A incrementing number that represents the number of unique nodes in the graph.
+        colors_: List[str]. A list for colors for nodes in the dotty graph.
+    """
+
     def __init__(self):
-        self.vertices_ = []
-        self.edges_ = []
-        self.uniqueVertexMap = {}
-        self.uniqueVertexNo = 0
-        self.colors = ["AliceBlue", "CadetBlue1", "Coral", "DarkOliveGreen1", "DarkSeaGreen1", "GhostWhite", "Khaki1", "LavenderBlush1", "LemonChiffon1",
-                       "LightSkyBlue", "MistyRose1", "MistyRose2", "PaleTurquoise2", "PeachPuff1", "PowderBlue", "Salmon", "Thistle1", "Thistle3", "Wheat1", "Yellow2"]
-
-    def get_unique_vertex_name(self, node):
-        if node not in self.uniqueVertexMap:
-            self.uniqueVertexMap[node] = self.uniqueVertexNo
-            self.uniqueVertexNo += 1
-        return "v" + str(self.uniqueVertexMap[node])
-
-    def dump_label(self, node):
-        labelStr = "{"
-        labelStr += " {<Inputs>Inputs}|"
-        labelStr += "{" + node.get_kind_name() + "\lname: " + \
-            node.get_name() + "}|"
-        labelStr += " {<Outputs>Outputs}"
-        labelStr += "}"
+        self.vertices_: List[str] = []
+        self.edges_: List[str] = []
+        self.uniqueVertexMap_: Dict[Node, int] = {}
+        self.uniqueVertexNo_: int = 0
+        self.colors_: List[str] = [
+            "AliceBlue",
+            "CadetBlue1",
+            "Coral",
+            "DarkOliveGreen1",
+            "DarkSeaGreen1",
+            "GhostWhite",
+            "Khaki1",
+            "LavenderBlush1",
+            "LemonChiffon1",
+            "LightSkyBlue",
+            "MistyRose1",
+            "MistyRose2",
+            "PaleTurquoise2",
+            "PeachPuff1",
+            "PowderBlue",
+            "Salmon",
+            "Thistle1",
+            "Thistle3",
+            "Wheat1",
+            "Yellow2"]
```

Comment:
Nit: End this like with a comma and put the closing brace on the next line, e.g.:
```
letters: List[str] = [
  "a",
  "b",
  "c",
  ...
  "z",
]
```

Reply:
This way, only one line needs to be modified to add a new list item.